taichi/transforms/bit_loop_vectorize.cpp

```diff
@@ -155,17 +156,180 @@ class BitLoopVectorize : public IRVisitor {
+          if (auto ptr = lhs->ptr->cast<GlobalPtrStmt>();
+              ptr && ptr->is_bit_vectorized) {
+            int32 rhs_val = get_constant_value(stmt->rhs);
+            // TODO: we limit 1 for now, 0 should be easy to implement by a
+            // bit_not on original bit pattern
+            TI_ASSERT(rhs_val == 1);
+            // cmp_eq with 1 yields the bit pattern itself
+
+            // to pass CFG analysis and mark the stmt vectorized
+            // create a dummy lhs + 0 here
+            auto zero = std::make_unique<ConstStmt>(TypedConstant(0));
+            auto add = std::make_unique<BinaryOpStmt>(BinaryOpType::add,
+                                                      stmt->lhs, zero.get());
+            add->is_bit_vectorized = true;
+            // modify IR
+            auto zero_p = zero.get();
+            stmt->insert_before_me(std::move(zero));
+            stmt->replace_with(add.get());
+            zero_p->insert_after_me(std::move(add));
+          }
+        } else if (auto lhs = stmt->lhs->cast<LocalLoadStmt>()) {
+          // case 1: lhs is a local load from a local adder structure
+          auto it = transformed_atomics.find(lhs->ptr[0].var);
+          if (it != transformed_atomics.end()) {
+            int32 rhs_val = get_constant_value(stmt->rhs);
+            // TODO: we limit 2 and 3 for now, the other case should be
+            // implement in a similar fashion
+            TI_ASSERT(rhs_val == 2 || rhs_val == 3);
+            // 010 and 011 respectively
+            auto &buffer_vec = it->second;
+            Stmt *a = buffer_vec[0], *b = buffer_vec[1], *c = buffer_vec[2];
+            // load all three buffers
+            auto load_a = std::make_unique<LocalLoadStmt>(LocalAddress(a, 0));
+            auto load_b = std::make_unique<LocalLoadStmt>(LocalAddress(b, 0));
+            auto load_c = std::make_unique<LocalLoadStmt>(LocalAddress(c, 0));
+            // compute not_a first
+            auto not_a = std::make_unique<UnaryOpStmt>(UnaryOpType::bit_not,
+                                                       load_a.get());
+            // b should always be itself so do nothing
+            // compute not_c
+            auto not_c = std::make_unique<UnaryOpStmt>(UnaryOpType::bit_not,
+                                                       load_c.get());
+            // bit_and all three patterns
+            auto and_a_b = std::make_unique<BinaryOpStmt>(
+                BinaryOpType::bit_and, not_a.get(), load_b.get());
+            auto and_b_c = std::make_unique<BinaryOpStmt>(
+                BinaryOpType::bit_and, and_a_b.get(),
+                rhs_val == 2 ? (Stmt *)(not_c.get()) : (Stmt *)(load_c.get()));
+            // mark the last stmt as vectorized
+            and_b_c->is_bit_vectorized = true;
+            // modify IR
+            auto and_a_b_p = and_a_b.get();
+            stmt->insert_before_me(std::move(load_a));
```

Comment:
A more elegant way to do this: https://github.com/taichi-dev/taichi/blob/b13abfaf8cb7c8b7a03d03ec4422795bf50dcda6/taichi/transforms/demote_operations.cpp#L22

Use `VecStatement::push_back<...>` to create the statements and `stmt->insert_before_me(vec_stmt)` to insert. (You don't need the `DelayedIRModifier` part.)

Reply:
I find to do it the elegant way we may still need `DelayedIRModifier` and there are other places that require changes as well(e.g. see the visitor for `GlobalLoadStmt`), therefore I think we should do it in a separate PR later and refactoring all this together. This also applies to the changes for `BinaryOpStmt::is_bit_vectorized`.